platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/moduleconverter/dependencies/DefaultLocalConfigurationMetadataBuilder.java

```diff
@@ -88,11 +87,17 @@ public LocalConfigurationMetadata create(
         ModelContainer<?> model,
         CalculatedValueContainerFactory calculatedValueContainerFactory
     ) {
+        // Ensure all actions are executed against this configuration and its hierarchy before reading its state.
+        // Dependency actions may modify the hierarchy.
+        runDependencyActionsInHierarchy(configuration);
```

Comment:
What's the implication of this running before we visit the variants below?

I think it slightly changes some evaluation ordering that would break certain side-effecty code. e.g., as it works today, if `collectVariants` causes user code to run that adds a configuration to `extendsFrom`, it's honored. But with the re-ordering, it will not.

Reply:
We discussed and we're going to leave this as-is. We need to move this up here in order to ensure the hierarchy is correct later, as many of the operations in this file iterate over the hierarchy. Really, the dependency actions is the only callback we really _support_ modifying extendsFrom. 

Eventually IMO we should deprecate changing `extendsFrom` after this method is called.